src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToCodeBehindCodeActionProvider.cs

```diff
@@ -52,14 +52,14 @@ public ExtractToCodeBehindCodeActionProvider(ILoggerFactory loggerFactory)
             return s_emptyResult;
         }
 
-        var change = new SourceChange(context.Location.AbsoluteIndex, length: 0, newText: string.Empty);
         var syntaxTree = context.CodeDocument.GetSyntaxTree();
         if (syntaxTree?.Root is null)
         {
             return s_emptyResult;
         }
 
-        var owner = syntaxTree.Root.LocateOwner(change);
+        var token = syntaxTree.Root.FindToken(context.Location.AbsoluteIndex);
+        var owner = token?.Parent;
```

Comment:
Can you validate that the switch on line 69 below, is still needed? Some of those quirks in tree shape might have been fixed by the `FindToken` API.

Reply:
It looks like it's still needed. Handle_AtEndOfCodeDirectiveWithNoSpace_ReturnsResult fails for 

```razor
@code$${ private var x = 1; }
```

Reply:
Thanks for checking. I suspect `@code$$ {` falls into the second arm now, not the first, but makes sense that the first one is needed with the current algorithm.

@333fred does Roslyn do any heuristics like "if the token in front of the caret is punctuation, use the token behind it"?

Reply:
Nope. The rule is very simple and straightforward: all trailing trivia, up to and not including the newline, is attached to the previous token. Any further trivia is attached to the next token (which is potentially the eof token at the end of the file).

Reply:
Thanks for confirming. If this becomes important we can always create an extension method that deals with it. As it is the switch expression only deals with two cases, so its not quite at the level of "deal with compiler quirks" :)